lib/eventFactory.js

```diff
@@ -453,50 +455,88 @@ async function convertStages(config) {
             }
         });
 
+        newStage.eventId = eventId;
         newStage.jobIds = jobIds;
+
         delete newStage.jobs; // extra field from yaml parser
 
+        // Check for setup and teardown
+        const setupJobName = `${STAGE_PREFIX}${key}:setup`;
+        const teardownJobName = `${STAGE_PREFIX}${key}:teardown`;
+        const setupJobExists = Object.keys(jobs).find(j => j === setupJobName || j === `~${setupJobName}`);
+        const teardownJobExists = Object.keys(jobs).find(j => j === teardownJobName || j === `~${teardownJobName}`);
+
+        if (setupJobExists) {
+            newStage.setup = [pipelineJobs.find(j => j.name === setupJobExists).id];
+        }
+        if (teardownJobExists) {
+            newStage.teardown = [pipelineJobs.find(j => j.name === teardownJobExists).id];
+        }
+
+        // Add workflow graph
         newStages.push(newStage);
     });
 
     return newStages;
 }
 
+/**
+ * Create stage model with stage metadata
+ * Create instance of a stage with stageId, eventId, workflowGraph, and status
+ * @method createStageBuilds
+ * @param  {Object}     config              config
+ * @param  {Number}     config.eventId      Event ID
+ * @param  {Object}     config.pipeline     Pipeline
+ * @return {Promise}
+ */
+async function createStageBuilds({ stageFactory, stageBuildFactory, stageConfig }) {
+    const stageCreateConfig = stageConfig;
+    const stageBuildCreateConfig = {
+        eventId: stageConfig.eventId,
+        workflowGraph: stageConfig.workflowGraph,
+        status: 'CREATED'
+    };
+
+    delete stageCreateConfig.eventId;
+    delete stageCreateConfig.workflowGraph;
+
+    stageCreateConfig.archived = false;
+
+    const stage = await stageFactory.create(stageCreateConfig);
+
+    stageBuildCreateConfig.stageId = stage.id;
+    await stageBuildFactory.create(stageBuildCreateConfig);
+
+    return stage;
+}
+
 /**
  * Sync stages
  * 1. Convert new stages into correct format, prepopulate with jobIds
  * 2. Create the stages if it is defined and was not in the database yet
  * @method createStages
  * @param  {Object}     config              config
- * @param  {Number}     config.groupEventId Group event ID
+ * @param  {Number}     config.eventId      Event ID
  * @param  {Object}     config.pipeline     Pipeline
  * @return {Promise}
  */
-async function createStages({ pipeline, groupEventId }) {
+async function createStages({ pipeline, eventId }) {
```

Comment:
Similar to rest of the pipeline configuration, shouldn't creating/updating stages be performed as part of pipeline sync flow?
https://github.com/screwdriver-cd/models/blob/master/lib/pipeline.js#LL864

Is there a reason why it is being deferred until an event is created?

Reply:
thought it would be more straightforward since stageBuilds need to be created during event creation and we need to get the stage IDs and workflowGraphs to create stageBuilds
I can move stage creation to pipeline sync